Keep trying other methods when transfer_method = smart

albertomurillo opened the pull request:
##### SUMMARY
Commit 1fe67f9 introduced a change to keep retrying file
transfers where previously retry only applied to commands.

This change broke transfer_method = smart in a way that
when sftp transfer is tried, it raises an exception
interrupting the flow to continue trying other methods.

This change catches AnsibleConnectionFailure in
_file_transport_command to continue trying other
methods.

Fixes #23711

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
plugins/connection/ssh.py

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.4.0 (devel 698fa37a44) last updated 2017/04/18 11:06:11 (GMT -500)
  config file = /Users/albertom/Desktop/test/ansible.cfg
  configured module search path = [u'/Users/albertom/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/albertom/github/ansible/lib/ansible
  executable location = /Users/albertom/github/ansible/bin/ansible
  python version = 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 12:39:47) [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
```

##### ADDITIONAL INFORMATION
before
```
# ansible-playbook -i hosts test.yml

PLAY [all] ***********************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************
fatal: [wolverine]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Connection closed\r\n", "unreachable": true}
	to retry, use: --limit @/Users/albertom/Desktop/test/test.retry

PLAY RECAP ***********************************************************************************************************************************************************************************************
wolverine                  : ok=0    changed=0    unreachable=1    failed=0

```

after
```
$ ansible-playbook -i hosts test.yml

PLAY [all] ***********************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************
 [WARNING]: sftp transfer mechanism failed on [wolverine]. Use ANSIBLE_DEBUG=1 to see detailed information

ok: [wolverine]

TASK [Test] **********************************************************************************************************************************************************************************************
ok: [wolverine] => {
    "msg": "This is a test"
}

PLAY RECAP ***********************************************************************************************************************************************************************************************
wolverine                  : ok=2    changed=0    unreachable=0    failed=0

```

gundalow commented:
Note to reviews: https://github.com/ansible/ansible/pull/23717/files?w=1

sivel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -801,30 +801,33 @@ def _file_transport_command(self, in_path, out_path, sftp_action):
-                    out_file.close()
-                else:
-                    in_data = open(to_bytes(in_path, errors='surrogate_or_strict'), 'rb').read()
+            try:
```
We try and wrap the smallest amount of code possible in a try/except.  Our preference would be to only wrap the `_run` and `exec_command`  lines in try/except.

Once this is done, we'll need to review again, but overall this approach looks correct.

albertomurillo replied:
Hi I updated the PR with the requested change

sivel commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -805,26 +805,37 @@ def _file_transport_command(self, in_path, out_path, sftp_action):
-                (returncode, stdout, stderr) = self._run(cmd, in_data, checkrc=False)
+                try:
+                    (returncode, stdout, stderr) = self._run(cmd, in_data, checkrc=False)
+                except AnsibleConnectionFailure:
```
You seem to be missing a `returncode = 255` here.  As such, this PR contains a SyntaxError.

albertomurillo replied:
Yea sorry about that. It is fixed now

albertomurillo replied:
@sivel can u re-review this one?

bcoca commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -805,26 +805,38 @@ def _file_transport_command(self, in_path, out_path, sftp_action):
+                try:
+                    (returncode, stdout, stderr) = self._run(cmd, in_data, checkrc=False)
+                except AnsibleConnectionFailure:
+                    returncode = 255
```
why not just push this into the _run method instead of adding to almost every invocation of it?

sivel replied:
`_ssh_retry` is at fault here, as it wraps `_run`.  And `_ssh_retry` raises `AnsibleConnectionFailure`.  We can see if it is feasible to have `_ssh_retry` return the "return tuple" if it has reached it's max attempts.  I don't know how much that would affect the rest of the code though.

The current approach is pretty targeted.

bcoca replied:
just looks as a lot of duplication, seems simpler to set a flag return 255 or exception depending on the flag

sivel commented:
@albertomurillo to follow up on the question @bcoca had, does the following resolve it instead of the approach of the multiple `try/except` blocks?

```diff
diff --git a/lib/ansible/plugins/connection/ssh.py b/lib/ansible/plugins/connection/ssh.py
index 12bdaf5..5701159 100644
--- a/lib/ansible/plugins/connection/ssh.py
+++ b/lib/ansible/plugins/connection/ssh.py
@@ -169,7 +169,10 @@ def _ssh_retry(func):
                     raise AnsibleConnectionFailure("Failed to connect to the host via ssh: %s" % to_native(return_tuple[2]))
             except (AnsibleConnectionFailure, Exception) as e:
                 if attempt == remaining_tries - 1:
-                    raise
+                    if isinstance(e, AnsibleConnectionFailure):
+                        return return_tuple
+                    else
+                        raise e
                 else:
                     pause = 2 ** attempt - 1
                     if pause > 30:
```

albertomurillo commented:
@sivel yes it works but there are some issue with it.

If i set ANSIBLE_SSH_RETRIES=4, it will try 4 times to run a command with sftp and until then, it will try scp, then the next task, will try 4 times with sftp and then try with scp and so on.

sivel commented:
@albertomurillo Ideally, that is expected.  In fact, regardless of which approach you use, it should work the same.  In either solution, either the `return` or the `raise` would not happen until the retries have been completed.

If you could, please implement my new recommendation.

Thanks!

Commit:
Keep trying other methods when transfer_method = smart

Commit 1fe67f9 introduced a change to keep retrying file
transfers where previously retry only applied to commands.

This change broke transfer_method = smart in a way that
when sftp transfer is tried, it raises an exception
interrupting the flow to continue trying other methods.

This commit changes _ssh_retry function to return the
return_tuple instead of raising an Exception.

Fixes #23711

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/20119/2/tests

```
>                       return return_tuple
E                       UnboundLocalError: local variable 'return_tuple' referenced before assignment

lib/ansible/plugins/connection/ssh.py:173: UnboundLocalError
```

abadger commented:
@sivel, so I'm thinking about @albertomurillo's objection and I think his point is valid.  Let's say you're on vacation in some Mediterranean Paradise with really flakey internet access.  You get a call from the big boss that something is really broken, could you please interrupt your vacation to fix it?  You figure out a fix and run an ansible command from your laptop to hit 100 heterogeneous hosts with that fix.  Because of the flakey internet, you have retries set to 10.  You also have transport set to smart because some of these hosts have just scp enabled, others have just sftp, and a few have just ssh.

For the hosts that will only work with piped, do you really want to wait for 20 failures before the first chance at a successful connection?

Further I think that transport = smart could very well mean that you should try each of the transports in turn , if all fail, count that as one retry, and try each of the transports a second time.  The transport type is the smallest atom of code that the retries should be acting upon, not any sub-transports that the transport invokes.
